### PR TITLE
Fix bootstarp errors

### PR DESCRIPTION
**Pull request by vancexu**
https://github.com/uber/cadence/issues/2617

### CODE REVIEW

File: service/worker/service.go [hunk 1]
```diff
@@ -327,17 +331,38 @@ func (s *Service) ensureSystemDomainExists(pFactory persistencefactory.Factory,
...
-	case *shared.EntityNotExistsError:
-		s.logger.Info("cadence-system domain does not exist, attempting to register domain")
-		s.registerSystemDomain(pFactory, clusterName)
-	default:
+	if err == nil {
+		s.ensureDomainAvailable()
+	} else {
+		if _, ok := err.(*shared.EntityNotExistsError); ok {
+			s.logger.Info("cadence-system domain does not exist, attempting to register domain")
+			s.registerSystemDomain(pFactory, clusterName)
```
Comment (yycptt) [inline service/worker/service.go]: Ensure domain available after registration?

File: service/worker/service.go [hunk 2]
```diff
@@ -327,17 +331,38 @@ func (s *Service) ensureSystemDomainExists(pFactory persistencefactory.Factory,
...
 		s.logger.Fatal("failed to verify if cadence system domain exists", tag.Error(err))
 	}
 }
 
+func (s *Service) ensureDomainAvailable() {
+	client := cclient.NewClient(s.params.PublicClient, common.SystemLocalDomainName, &cclient.Options{})
+	err := client.TerminateWorkflow(context.Background(), "wid-not-exist", "", "test reason", nil)
+	retryCount := 0
+	for err != nil && retryCount <= 10 {
+		nonExistErr, ok := err.(*cshared.EntityNotExistsError)
```
Comment (yycptt) [inline service/worker/service.go]: What about other error types like service busy? Do we need to keep retrying in that case?

File: service/worker/service.go [hunk 3]
```diff
@@ -327,17 +331,38 @@ func (s *Service) ensureSystemDomainExists(pFactory persistencefactory.Factory,
...
+			s.logger.Info("cadence-system domain does not exist, attempting to register domain")
+			s.registerSystemDomain(pFactory, clusterName)
+		}
 		s.logger.Fatal("failed to verify if cadence system domain exists", tag.Error(err))
 	}
 }
 
+func (s *Service) ensureDomainAvailable() {
+	client := cclient.NewClient(s.params.PublicClient, common.SystemLocalDomainName, &cclient.Options{})
+	err := client.TerminateWorkflow(context.Background(), "wid-not-exist", "", "test reason", nil)
```
Comment (yycptt) [inline service/worker/service.go]: Consider adding a comment explaining why `TerminateWorkflow` is used here?

File: service/worker/service.go [hunk 4]
```diff
@@ -327,17 +331,38 @@ func (s *Service) ensureSystemDomainExists(pFactory persistencefactory.Factory,
...
+			err = client.TerminateWorkflow(context.Background(), "wid-not-exist", "", "test reason", nil)
+			retryCount++
+		} else {
+			break
+		}
+	}
+}
+
+func isErrSystemDomainNotExist(err *cshared.EntityNotExistsError) bool {
+	return err.Message == fmt.Sprintf("Domain: %s", common.SystemLocalDomainName)
```
Comment (yycptt) [inline service/worker/service.go]: Equality check seems a bit fragile. Is it possible to just check part of the message like `SystemLocalDomainName` or `Domain`?

File: service/worker/service.go (hunk 1 above)
Comment (vancexu) [reply service/worker/service.go]: logically we should but registerSystemDomain already wait for cache refresh, so avoid additional call here.

File: service/worker/service.go (hunk 3 above)
Comment (vancexu) [reply service/worker/service.go]: added

File: service/worker/service.go (hunk 2 above)
Comment (vancexu) [reply service/worker/service.go]: We should only wait on this DomainNotExist error. Retry on other error make little sense.

File: service/worker/service.go (hunk 4 above)
Comment (vancexu) [reply service/worker/service.go]: Addressed.

**Pull request merged**